File: js/confidential/src/verify/venice.ts
// Venice TEE/E2EE verifier. Venice returns an Intel TDX quote, NVIDIA evidence, a
// nonce, and a secp256k1 signing/encryption key. We validate the structural quote
// bindings: the signing address derives from the attested key (keccak256 Ethereum
// address), the address is bound in report_data[0:32], the caller nonce is bound in
// report_data[64:128], the model matches, the key is in the attested workload
// keyset, and the nested attestation report_data matches the quote. Full
// Intel/NVIDIA vendor-root verification is deliberately not inferred, so the honest
// ceiling is `provider-attested`.

import { assembleResult, check, freshnessCheck, hexEqual, readEnvelope } from "./checks.js";
import { secp256k1AddressFromPublicKey } from "./crypto.js";
import { parseTdxQuote, TDX_TEE_TYPE } from "./tdx.js";
import type {
  AttestationCheck,
  AttestationExpectations,
  NormalizedAttestationResult,
  TeeVerifier
} from "./types.js";

interface VeniceAttestationPayload {
  intel_quote?: unknown;
  nvidia_payload?: unknown;
  nonce?: unknown;
  model?: unknown;
  signing_address?: unknown;
  signing_public_key?: unknown;
  signing_key?: unknown;
  signing_algo?: unknown;
  attestation?: {
    report_data?: unknown;
    workload_keyset?: { e2ee_public_keys?: Array<{ algo?: unknown; public_key?: unknown }> };
    evidence?: { quote_report_data?: unknown };
  };
}

export interface VeniceVerifierOptions {
  verifierVersion?: string;
}

export class VeniceTeeVerifier implements TeeVerifier {
  readonly provider = "venice";
  readonly verifierVersion: string;

  constructor(opts: VeniceVerifierOptions = {}) {
    this.verifierVersion = opts.verifierVersion ?? "venice-tdx-receipt/2";
  }

  supportsClientOpaqueE2ee(): boolean {
    return true;
  }

  verifyAttestation(evidence: unknown, expectations: AttestationExpectations): NormalizedAttestationResult {
    const envelope = readEnvelope(evidence, expectations);
    const payload = envelope.payload as VeniceAttestationPayload | undefined;
    const checks: AttestationCheck[] = [];
    const quoteRaw = typeof payload?.intel_quote === "string" ? payload.intel_quote : null;
    const parsed = quoteRaw ? parseTdxQuote(quoteRaw) : null;
    const nonce = typeof payload?.nonce === "string" ? payload.nonce : null;
    const model = typeof payload?.model === "string" ? payload.model : null;
    const signingAddress = typeof payload?.signing_address === "string" ? payload.signing_address.toLowerCase() : null;
    const signingPublicKey = typeof payload?.signing_public_key === "string"
      ? payload.signing_public_key
      : typeof payload?.signing_key === "string" ? payload.signing_key : null;
    const derivedAddress = signingPublicKey ? secp256k1AddressFromPublicKey(signingPublicKey) : null;

    checks.push(check("evidence_present", Boolean(payload && quoteRaw), true, quoteRaw ? undefined : "no Intel quote"));
    checks.push(check("quote_parsed", parsed !== null, true, parsed ? undefined : "TDX quote did not parse"));
    checks.push(check("expected_tee_type", parsed?.teeType === TDX_TEE_TYPE, true,
      parsed?.teeType === TDX_TEE_TYPE ? undefined : "not an Intel TDX quote"));
    checks.push(check("debug_disabled", parsed !== null && !parsed.debugEnabled, true,
      parsed?.debugEnabled ? "TD debug mode enabled" : undefined));
    checks.push(check("nonce_binding", Boolean(parsed && nonce && hexEqual(nonce, expectations.nonce)
      && hexEqual(parsed.reportData.slice(64, 128), expectations.nonce)), true,
    nonce ? undefined : "attestation did not carry the caller nonce"));
    checks.push(check("model_binding", model === expectations.upstreamModel, true,
      model ? undefined : "attestation did not name the route model"));
    checks.push(check("signing_algorithm", payload?.signing_algo === "ecdsa", true,
      payload?.signing_algo === "ecdsa" ? undefined : "unsupported attested signing algorithm"));
    checks.push(check("signing_key_address", Boolean(derivedAddress && signingAddress && hexEqual(derivedAddress, signingAddress)), true,
      derivedAddress ? undefined : "attested secp256k1 key was missing or malformed"));

    const addressReportPrefix = signingAddress && /^0x[0-9a-f]{40}$/.test(signingAddress)
      ? signingAddress.slice(2).padEnd(64, "0")
      : null;
    checks.push(check("signing_address_quote_binding", Boolean(parsed && addressReportPrefix
      && hexEqual(parsed.reportData.slice(0, 64), addressReportPrefix)), true,
    addressReportPrefix ? undefined : "attested signing address was malformed"));

    // ABSENT IS NOT THE SAME AS WRONG, and the difference decides what anyone
    // does next. "Did not match" says the provider asserted two things that
    // contradict each other — an inconsistency to report to them. "Carried
    // nothing" says the provider asserted nothing at all — a route to withhold
    // until it does. Three live Venice routes take the second path today and
    // were being reported as the first, which sends a reader looking for a
    // mismatch that does not exist.
    //
    // The VERDICT is identical either way: a binding nobody stated is a binding
    // that did not hold. Only the reason changes.
    const nested = payload?.attestation;
    const nestedDocumentPresent = Boolean(nested) && typeof nested === "object"
      && Object.keys(nested as Record<string, unknown>).length > 0;
    const reportedReportData = typeof payload?.attestation?.report_data === "string" ? payload.attestation.report_data : null;
    const evidenceReportData = typeof payload?.attestation?.evidence?.quote_report_data === "string"
      ? payload.attestation.evidence.quote_report_data : null;
    const reportedQuoteBound = Boolean(parsed
      && hexEqual(reportedReportData, parsed.reportData)
      && hexEqual(evidenceReportData, parsed.reportData));
    checks.push(check("reported_quote_binding", reportedQuoteBound, true,
      reportedQuoteBound
        ? undefined
        : !nestedDocumentPresent
          ? "the evidence carried no nested attestation document, so nothing restates the quote's report_data"
          : reportedReportData === null || evidenceReportData === null
            ? "the nested attestation document omits report_data"
            : "nested attestation report_data did not match the quote"));

    const keysetKeys = payload?.attestation?.workload_keyset?.e2ee_public_keys;
    const keyInWorkload = Array.isArray(keysetKeys) && keysetKeys.some((entry) =>
      entry?.algo === "secp256k1-aes-256-gcm-hkdf-sha256"
      && typeof entry.public_key === "string"
      && signingPublicKey !== null
      && hexEqual(entry.public_key, signingPublicKey)
    );
    checks.push(check("workload_keyset_binding", keyInWorkload, true,
      keyInWorkload
        ? undefined
        : !Array.isArray(keysetKeys)
          ? "the evidence attests no workload keyset, so nothing binds the signing key to the workload that ran"
          : "the attested workload keyset does not contain the signing key"));

    const gpuPresent = typeof payload?.nvidia_payload === "string" && payload.nvidia_payload.length > 0;
    checks.push(check("gpu_evidence_present", gpuPresent, true, gpuPresent ? undefined : "no NVIDIA GPU evidence"));
    checks.push(freshnessCheck(envelope.fetchedAtMs, expectations));

    const measurements: Record<string, string> = parsed ? {
      mrtd: parsed.mrTd,
      mr_config_id: parsed.mrConfigId,
      rtmr0: parsed.rtmr0,
      rtmr1: parsed.rtmr1,
      rtmr2: parsed.rtmr2,
      rtmr3: parsed.rtmr3
    } : {};
    return assembleResult({
      expectations,
      hardwareType: "intel-tdx+nvidia-cc",
      requestedLevel: "provider-attested",
      privacyModality: expectations.privacyModality,
      measurementIdentities: measurements,
      modelWeightIdentity: null,
      attestedTlsSpki: null,
      attestedEncryptionKey: signingPublicKey,
      attestedSigningKey: signingAddress,
      boundNonce: parsed && nonce && hexEqual(nonce, expectations.nonce) ? expectations.nonce : nonce,
      verifierVersion: this.verifierVersion,
      supportsClientOpaqueE2ee: true,
      checks
    });
  }
}

File: python/src/anonrouter_confidential/verify/venice.py
"""Venice TEE/E2EE verifier.

Venice returns an Intel TDX quote, NVIDIA evidence, a nonce, and a secp256k1
signing/encryption key. We validate the structural quote bindings locally. Full
Intel/NVIDIA vendor-root verification is deliberately NOT inferred from Venice's
own ``verified`` booleans, so the honest maximum stays ``provider-attested``.

Verdict-for-verdict equivalent to the Venice verifier in ``@anonrouter/confidential``.
"""

from __future__ import annotations

import re
from typing import Any

from .._util import as_dict, as_str
from ..crypto.venice import eth_address_from_pub
from ..tdx import TDX_TEE_TYPE, parse_tdx_quote
from .checks import assemble_result, check, freshness_check, hex_equal, read_envelope
from .types import AttestationExpectations, NormalizedVerdict

VERIFIER_VERSION = "venice-tdx-receipt/2"
_VENICE_KEYSET_ALGO = "secp256k1-aes-256-gcm-hkdf-sha256"
_ADDR_RE = re.compile(r"^0x[0-9a-f]{40}$")


def _str_or_none(value: Any) -> str | None:
    return as_str(value)


def _derived_address(signing_public_key: str | None) -> str | None:
    if not signing_public_key:
        return None
    try:
        return eth_address_from_pub(signing_public_key)
    except ValueError:
        return None


def verify_venice(evidence: Any, expectations: AttestationExpectations) -> NormalizedVerdict:
    envelope = read_envelope(evidence, expectations)
    payload = as_dict(envelope.payload)
    checks = []

    quote_raw = _str_or_none(payload.get("intel_quote"))
    parsed = parse_tdx_quote(quote_raw) if quote_raw else None
    nonce = _str_or_none(payload.get("nonce"))
    model = _str_or_none(payload.get("model"))
    signing_address_raw = _str_or_none(payload.get("signing_address"))
    signing_address = signing_address_raw.lower() if signing_address_raw else None
    signing_public_key = _str_or_none(payload.get("signing_public_key")) or _str_or_none(payload.get("signing_key"))
    derived_address = _derived_address(signing_public_key)

    checks.append(check("evidence_present", bool(payload and quote_raw), True, None if quote_raw else "no Intel quote"))
    checks.append(check("quote_parsed", parsed is not None, True, None if parsed else "TDX quote did not parse"))
    checks.append(check(
        "expected_tee_type",
        parsed is not None and parsed.tee_type == TDX_TEE_TYPE,
        True,
        None if (parsed and parsed.tee_type == TDX_TEE_TYPE) else "not an Intel TDX quote",
    ))
    checks.append(check(
        "debug_disabled",
        parsed is not None and not parsed.debug_enabled,
        True,
        "TD debug mode enabled" if (parsed and parsed.debug_enabled) else None,
    ))
    checks.append(check(
        "nonce_binding",
        bool(parsed and nonce and hex_equal(nonce, expectations.nonce)
             and hex_equal(parsed.report_data[64:128], expectations.nonce)),
        True,
        None if nonce else "attestation did not carry the caller nonce",
    ))
    checks.append(check(
        "model_binding",
        model == expectations.upstream_model,
        True,
        None if model else "attestation did not name the route model",
    ))
    checks.append(check(
        "signing_algorithm",
        payload.get("signing_algo") == "ecdsa",
        True,
        None if payload.get("signing_algo") == "ecdsa" else "unsupported attested signing algorithm",
    ))
    checks.append(check(
        "signing_key_address",
        bool(derived_address and signing_address and hex_equal(derived_address, signing_address)),
        True,
        None if derived_address else "attested secp256k1 key was missing or malformed",
    ))

    address_report_prefix = (
        signing_address[2:].ljust(64, "0")
        if signing_address and _ADDR_RE.match(signing_address)
        else None
    )
    checks.append(check(
        "signing_address_quote_binding",
        bool(parsed and address_report_prefix and hex_equal(parsed.report_data[0:64], address_report_prefix)),
        True,
        None if address_report_prefix else "attested signing address was malformed",
    ))

    # ABSENT IS NOT THE SAME AS WRONG, and the difference decides what anyone
    # does next. "Did not match" says the provider asserted two contradictory
    # things -- an inconsistency to report to them. "Carried nothing" says the
    # provider asserted nothing at all -- a route to withhold until it does.
    # Three live Venice routes take the second path today and were being
    # reported as the first, which sends a reader hunting for a mismatch that
    # does not exist.
    #
    # The VERDICT is identical either way: a binding nobody stated is a binding
    # that did not hold. Only the reason changes.
    attestation = as_dict(payload.get("attestation"))
    nested_document_present = len(attestation) > 0
    reported_report_data = _str_or_none(attestation.get("report_data"))
    evidence_block = as_dict(attestation.get("evidence"))
    evidence_report_data = _str_or_none(evidence_block.get("quote_report_data"))
    reported_quote_bound = bool(
        parsed
        and hex_equal(reported_report_data, parsed.report_data)
        and hex_equal(evidence_report_data, parsed.report_data)
    )
    checks.append(check(
        "reported_quote_binding",
        reported_quote_bound,
        True,
        None
        if reported_quote_bound
        else "the evidence carried no nested attestation document, so nothing restates the quote's report_data"
        if not nested_document_present
        else "the nested attestation document omits report_data"
        if reported_report_data is None or evidence_report_data is None
        else "nested attestation report_data did not match the quote",
    ))

    keyset = as_dict(attestation.get("workload_keyset"))
    keyset_keys = keyset.get("e2ee_public_keys")
    key_in_workload = isinstance(keyset_keys, list) and any(
        isinstance(entry, dict)
        and entry.get("algo") == _VENICE_KEYSET_ALGO
        and isinstance(entry.get("public_key"), str)
        and signing_public_key is not None
        and hex_equal(entry["public_key"], signing_public_key)
        for entry in keyset_keys
    )
    checks.append(check(
        "workload_keyset_binding",
        key_in_workload,
        True,
        None
        if key_in_workload
        else "the evidence attests no workload keyset, so nothing binds the signing key to the workload that ran"
        if not isinstance(keyset_keys, list)
        else "the attested workload keyset does not contain the signing key",
    ))

    gpu = payload.get("nvidia_payload")
    gpu_present = isinstance(gpu, str) and len(gpu) > 0
    checks.append(check("gpu_evidence_present", gpu_present, True, None if gpu_present else "no NVIDIA GPU evidence"))
    checks.append(freshness_check(envelope.fetched_at_ms, expectations))

    measurements = (
        {
            "mrtd": parsed.mr_td,
            "mr_config_id": parsed.mr_config_id,
            "rtmr0": parsed.rtmr0,
            "rtmr1": parsed.rtmr1,
            "rtmr2": parsed.rtmr2,
            "rtmr3": parsed.rtmr3,
        }
        if parsed
        else {}
    )
    bound_nonce = expectations.nonce if (parsed and nonce and hex_equal(nonce, expectations.nonce)) else nonce
    return assemble_result(
        expectations=expectations,
        hardware_type="intel-tdx+nvidia-cc",
        requested_level="provider-attested",
        measurement_identities=measurements,
        model_weight_identity=None,
        attested_tls_spki=None,
        attested_encryption_key=signing_public_key,
        attested_signing_key=signing_address,
        bound_nonce=bound_nonce,
        verifier_version=VERIFIER_VERSION,
        supports_client_opaque_e2ee=True,
        checks=checks,
    )
